### ISSUE TITLE
checksums generate OK but install copies them with wrong target name, overlaying multiples

### ISSUE THREAD

**Issue by mischor** [nicoulaj/checksum-maven-plugin#63]
Configuring with goal "files", and using <attachChecksums>true</attachChecksums> generates the correct checksums with files properly named, but a subsequent maven-install-plugin install step copies the created checksum files in to the .m2 repository with the incorrect names.  Example:

[INFO] Installing
   ...
\parent-pom\target\parent-pom-12-SNAPSHOT-source-release.zip.sha512
to ...\.m2\repository\...\parent-pom\12-SNAPSHOT\parent-pom-12-SNAPSHOT.sha512
[INFO] Installing
   ...                        \parent-pom\target\parent-pom-12-SNAPSHOT.pom.sha512
to ...\.m2\repository\...\parent-pom\12-SNAPSHOT\parent-pom-12-SNAPSHOT.sha512

In summary, (assuming checksum is SHA512) install copies attached artifacts of name: 
<artifact-name>-<version>-<suffix>.<type>.sha512 to 
<project-name>-<version>.sha512  which is the same name for multiple checksummed artifacts.

**Comment by mischor** [comment]
I did a test: using "files" with an artifact which included a classifier: e.g. my-proj-2.10.3-SNAPSHOT-source-release.zip.

This fails because the creation of the ChecksumFile instance passes in "null" for the classifier.  If I override that in the source code (e.g., look in the file path to see if has "source-release" and if so, pass that as the classifier, it succeeds in attaching the checksum properly.

I think a proper fix for this will need either:
1) a reliable maven-endorsed way to extract a "type" and a "classifier" from a filepath, (which might not be doable, in general), or

2) an augmentation of the configuration of this plugin which adds to the <include> of the <fileSet> attributes for type and classifier, or 

3) a combination of the two (use the specified type/classifier if supplied, otherwise attempt to deduce them from the filePath)

Is this something that can be fixed quickly?  If not, since this is blocking our releases, we'll have to drop using this approach and go back to something like ANT scripts.

**Comment by mischor** [comment]
Just FYI: the "artifacts" goal also has an usual issue.  It seems to pick up other artifacts, but misses the pom itself.  I tried using the build-helper-maven-plugin to "attach" the pom but got an error indicating it was already "attached".  

One thing I've noticed that's different about the pom:  the source is not in the /target but is at the top level and is named "pom.xml".  The attached artifact is never put into the target (normally), but does appear in the .m2 repo, where it's renamed <artifact-id>-<version>.pom  (no xml, and using the standard maven naming conventions).

**Comment by mischor** [comment]
The recently (August 18th) updated  "apache-wide" super-pom ( https://search.maven.org/artifact/org.apache/apache/21/pom ) now uses this plugin, with the "files" option, attempting to add signatures for the source-release classifier artifacts.

**Comment by mischor** [comment]
The maven-gpg-plugin also has a "sign all the artifacts" option.  The code for this is in GpgSignAttachedMojo, and special-cases the pom separately from the other "attached" artifacts, and seems to have code for handling the case where the pom is not in the /target.  

I did notice at one spot it has a wrong value due to the assumption that the pom for deploy has a name made from ${finalName} - that is I think not correct, I think it is forced to be ${artifactId}-${version}.  This is, of course, the "default for finalName, but some users override this.  For things going into /target, this is allowed, but for things going into .m2 repositories, the official maven naming convention is (I believe) "enforced".  The line in the class (maybe) needing a fix would be this one:
File pomToSign = new File( project.getBuild().getDirectory(), project.getBuild().getFinalName() + ".pom" );

**Comment by mischor** [comment]
clarification: I ran a test using a pom with a finalName which was just ${artifactId}.  The maven-gpg-plugin created in /target a copy of the pom.xml file, with thename ${artifactId}.pom and signed it, creating in the /target a file ${artifactId}.pom.asc.   

The maven install plugin seems to have the code to force maven naming conventions - it changed the name of this file in the .m2 spot to ${artifactId}-${version}.pom, and the signature likewise to ${artifactId}-${version}.pom.asc.

**Comment by mischor** [comment]
If both this plugin and the maven-gpg-plugin are run, they should account for the possibility that the other plugin may have already copied the pom.xml into the target.

**Comment by mischor** [comment]
What seems to work is to have the ArtifactAttacher method artifactCreated call the attachArtifact with an attachementType that is made from the actual file name (which ends ....xxx.sha512) set to xxx.sha512.  Example: for the artifact produced by the maven-javadocs plugin "jar" goal (name = 
${finalName}-javadocs.jar), the checksum artifact is 
${finalName}-javadocs.jar.sha512 (this is the name in the /target dir) and the deployed name would be
${finalName}-javadocs.jar.sha512 (the attachArtifact attachementType = jar.sha512, the artifactClassifier is javadocs).
Here's a code snippet, without enough checking, which worked for me:
```java
    public void artifactCreated(File artifact, String checksumType, String artifactType, String artifactClassifier) {
        if (checksumType.startsWith(".")) {
            // Project helper expects a type without leading dot (e.g. turn ".md5" into "md5").
            checksumType = checksumType.substring(1);
        }

        String fileName = artifact.getName();
        int period1 = fileName.lastIndexOf('.');  // finds xxx.yyy.sha512   period in front of sha512
        int period2 = fileName.lastIndexOf('.', period1 - 1);
        String attachmentType = fileName.substring(period2 + 1);   // yyy.sha512

        System.out.println("debug artifactType: " + ((artifactType == null) ? "null" : artifactType));
        System.out.println("debug artifactClassifier: " + artifactClassifier);
        System.out.println("debug checksumType: " + ((checksumType == null) ? "null" : checksumType));
        System.out.println("debug derived attachmentType: " + attachmentType);
        System.out.println("debug artifact: " + artifact.getAbsolutePath());

        projectHelper.attachArtifact(project, attachmentType, artifactClassifier, artifact);
    }
```

**Comment by mischor** [comment]
here's a patch file fixing 2 items:  1) generating for the "artifacts" goal the proper names for artifacts where the artifact 'type' is different from the file-type, and 2) adding the "pom" to the list of things to generate checksums for, when using the artifacts goal.  
[checksum-maven-plugin-1.8-issue-64.txt](https://github.com/nicoulaj/checksum-maven-plugin/files/2457640/checksum-maven-plugin-1.8-issue-64.txt)

I'm hoping this fix can be released soon - it's kind of blocking our progress.

**Comment by nicoulaj** [comment]
Can you submit your patch as a pull request ? This way we can have the tests run and discuss changes.

It looks like the patch contains several unnecessary changes, like commented out debug traces and modifications on license headers. Also, looking at your code snippet I doubt looking for two dots in a filename is a robust solution, and I believe a deeper refactoring is needed...

**Comment by kwin** [comment]
@nicoulaj I provided a new PR for this issue at #87. Please have a look and let me know what you think.
